src/periphery/Rewarder.sol

```diff
@@ -0,0 +1,201 @@
+
+    /**
+     * @notice Initialize a new mSpellStaking contract
+     * @dev This contract needs to receive an IERC20 `_rewardToken` in order to distribute them
+     * (with MoneyMaker in our case)
+     * @param _mim The address of the MIM token
+     */
+    constructor(
+        IERC20 _mim,
+        ICauldronV4 _cauldron
+    ) {
+        require(address(_mim) != address(0), "mSpellStaking: reward token can't be address(0)");
+
+        mim = _mim;
+        cauldron = _cauldron;
+        degenBox = IBentoBoxV1(_cauldron.bentoBox());
+    }
+
+    modifier onlyCauldron {
+        require(msg.sender == address(cauldron), "Caller needs to be Cauldron");
+        _;
+    }
+
+    /**
+     * @notice Deposit SPELL for reward token allocation
+     * @param _amount The amount of SPELL to deposit
+     */
+    function deposit(uint256 _amount, address from) external onlyCauldron {
+        UserInfo storage user = userInfo[from];
+
+        user.amount = user.amount + _amount;
+
+        updateReward();
+
+        user.rewardDebt = user.rewardDebt + int256(_amount * accRewardPerShare / ACC_REWARD_PER_SHARE_PRECISION);
+
+        emit Deposit(from, _amount);
+    }
+
+    /**
+     * @notice View function to see pending reward token on frontend
+     * @param _user The address of the user
+     * @return `_user`'s pending reward token
+     */
+    function pendingReward(address _user) external view returns (uint256) {
+        UserInfo storage user = userInfo[_user];
+        uint256 _totalSpell = cauldron.totalCollateralShare();
+        uint256 _accRewardTokenPerShare = accRewardPerShare;
+
+        uint256 _rewardBalance = mim.balanceOf(address(this));
+
+        if (_rewardBalance != lastRewardBalance && _totalSpell != 0) {
+            uint256 _accruedReward = _rewardBalance - lastRewardBalance;
+            _accRewardTokenPerShare = _accRewardTokenPerShare + _accruedReward * ACC_REWARD_PER_SHARE_PRECISION / _totalSpell;
+        }
+        return uint256(int256(user.amount * _accRewardTokenPerShare / ACC_REWARD_PER_SHARE_PRECISION) - user.rewardDebt);
+    }
+
+    function _repayLoan(address to, uint256 amount, uint256 part) internal {
+        mim.safeTransfer(address(degenBox), amount + 1);
+        degenBox.deposit(mim, address(degenBox), address(degenBox), amount + 1, 0);
+        cauldron.repay(to, true, part);
+    }
+
+    function _harvest(UserInfo memory user, address to) internal returns (UserInfo memory) {
+        int256 accumulatedMim = int256(user.amount * accRewardPerShare / ACC_REWARD_PER_SHARE_PRECISION);
+        uint256 pendingRewards = uint256(accumulatedMim - user.rewardDebt) - 1;
+
+        user.rewardDebt = accumulatedMim;
+
+        if (pendingRewards != 0) {
+            uint256 borrowPart = cauldron.userBorrowPart(to);
+            Rebase memory totalBorrow = cauldron.totalBorrow();
+            uint256 elastic = totalBorrow.toElastic(borrowPart, true);
```

Comment:
So, it's different from `repayForAll` here we limit rewards to the borrow amount + debt. If you don't borrow anything, you yields are going to be allocated for user that actually borrow. So the more you borrow the more you get from your collateralShare yields

Reply:
here either your position is repaid or you get it back to your wallet

Reply:
oh right